x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java

```diff
@@ -197,8 +206,65 @@ private void innerExecute(Task task, EsqlQueryRequest request, ActionListener<Es
             remoteClusterService,
             planRunner,
             queryBuilderResolver,
-            listener.map(result -> toResponse(task, request, configuration, result))
+            ActionListener.wrap(result -> {
+                recordCCSTelemetry(task, executionInfo, request, null);
+                listener.onResponse(toResponse(task, request, configuration, result));
+            }, ex -> {
+                recordCCSTelemetry(task, executionInfo, request, ex);
+                listener.onFailure(ex);
+            })
         );
+
+    }
+
+    private void recordCCSTelemetry(Task task, EsqlExecutionInfo executionInfo, EsqlQueryRequest request, @Nullable Exception exception) {
+        if (executionInfo.isCrossClusterSearch() == false) {
+            return;
+        }
+
+        CCSUsage.Builder usageBuilder = new CCSUsage.Builder();
+        usageBuilder.setClientFromTask(task);
+        if (exception != null) {
```

Comment:
Should we also check for `isRemoteUnavailableException` and classify those? Maybe that's done elsewhere?

Reply:
`setFailure` will classify those if `classifyVerificationException` didn't.

Reply:
I didn't put the whole logic there because VerificationException is specific to ES|QL and I didn't want to put ES|QL stuff into generic code.